collab-document/src/blocks/block.rs

```diff
@@ -1,158 +1,160 @@
-    &self.root
+    let txn = self.root.transact();
+    let mut map = serializer.serialize_map(Some(self.root.len(&txn) as usize))?;
+    for (k, _) in self.root.iter(&txn) {
+      let block = self.get_block(&txn, k).unwrap();
+      let value = serde_json::to_value(block).unwrap();
+      map.serialize_entry(k, &value)?;
+    }
+    map.end()
   }
 }
 
-pub struct BlockBuilder<'a, 'b> {
-  block_map: BlockMapRef,
-  txn: &'a mut TransactionMut<'b>,
-}
-
-impl<'a, 'b> BlockBuilder<'a, 'b> {
-  pub fn new(txn: &'a mut TransactionMut<'b>, container: &MapRefWrapper) -> Self {
-    let key = nanoid!(4);
-    Self::new_with_txn(txn, key, container)
+impl BlockMap {
+  pub fn new(root: MapRefWrapper) -> Self {
+    Self { root }
   }
 
-  pub fn new_with_txn(
-    txn: &'a mut TransactionMut<'b>,
-    block_id: String,
-    container: &MapRefWrapper,
-  ) -> Self {
-    let map_ref = match container.get_map_with_txn(txn, &block_id) {
-      None => container.insert_map_with_txn(txn, &block_id),
-      Some(map) => map,
-    };
-    let block_map = BlockMapRef::from_map_ref(map_ref);
-
-    Self { block_map, txn }
+  pub fn to_json(&self) -> serde_json::Value {
+    serde_json::to_value(self).unwrap()
   }
 
-  pub fn with_type<T: AsRef<str>>(mut self, ty: T) -> Self {
-    self.block_map.set_ty(self.txn, ty.as_ref().to_string());
-    self
+  pub fn get_block<T: ReadTxn>(&self, txn: &T, block_id: &str) -> Option<Block> {
+    let block_map = self.root.get_map_with_txn(txn, block_id);
+    match block_map {
+      Some(block_map) => {
+        let block = self.get_block_by_map(txn, block_map);
+        Some(block)
+      },
+      None => None,
+    }
   }
 
-  pub fn with_data<T: ToString>(mut self, data: T) -> Self {
-    self.block_map.set_data(self.txn, data.to_string());
-    self
+  pub fn get_block_by_map<T: ReadTxn>(&self, txn: &T, block_map: MapRefWrapper) -> Block {
+    let id = block_map.get(txn, "id").unwrap().to_string(txn);
+    let ty = block_map.get(txn, "ty").unwrap().to_string(txn);
+    let parent = block_map.get(txn, "parent").unwrap().to_string(txn);
+    let children = block_map.get(txn, "children").unwrap().to_string(txn);
+    let data = block_map.get(txn, "data").unwrap().to_string(txn);
+    Block {
+      id,
+      ty,
+      parent,
+      children,
+      data,
+    }
   }
 
-  pub fn with_next<T: AsRef<str>>(mut self, next: T) -> Self {
-    self.block_map.set_next(self.txn, next.as_ref().to_string());
-    self
+  pub fn create_block(
+    &self,
+    txn: &mut TransactionMut,
+    block_id: String,
+    ty: String,
+    parent_id: String,
+    children_id: String,
+    data: BlockDataEnum,
+  ) -> Result<Block> {
+    let block = Block {
+      id: block_id.clone(),
+      ty,
+      parent: parent_id,
+      children: children_id,
+      data: data.to_string(),
+    };
+    let block_map = self.root.insert_map_with_txn(txn, &block_id);
+    block_map.insert_with_txn(txn, "id", block.id.clone());
```

Comment:
I think it would be better if these keys defines as const str. Because I noticed these are used not only one place. For example:

```rust
const BLOCK_ID: &str = "id";
```